src/mds/Locker.cc

```diff
@@ -4794,12 +4829,113 @@ void Locker::scatter_tick()
 	       << *lock << " " << *lock->get_parent() << dendl;
       continue;
     }
-    if (now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
+
+    if (!mdr && now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
       break;
     updated_scatterlocks.pop_front();
-    scatter_nudge(lock, 0);
+
+    dout(20) << __func__ << " lock: " << *lock << " of " << *lock->get_parent()
+      << ", update_stamp: " << lock->get_update_stamp() << ", lock's parent's rstat_dirty_from:"
+      << static_cast<CInode*>(lock->get_parent())->get_rstat_dirty_from() << ", lock's parent's subdir rstat_dirty_from:"
+      << static_cast<CInode*>(lock->get_parent())->get_dirfrags_rstat_dirty_from()
+      << ", mdr's rstats_propagate_time: " << rstat_propagate_time << dendl;
+
+    if (lock->get_type() == CEPH_LOCK_INEST) {
+
+      CInode* ino = static_cast<CInode*>(lock->get_parent());
+      if (lock->is_nudging()) {
+        // We need to keep an eye on nestlocks that are of the current mdr's interest
+        // and are already nudged by scatter_tick or other rstat flush operation(Although
+        // there can be at most only one client rstat flush request at any time, there
+        // could be multiple slave requests at the same time if the current mds is auth
+        // for multiple disjoint subtrees).
+        if (mdr && ino->need_to_nudge(rstat_propagate_time)
+            && !lock->is_nudged_by(mdr->reqid)) {
+
+          dout(20) << __func__ << " " << mdr->reqid << " nestlock of " << *lock->get_parent()
+            << " already nudged, adding callback" << dendl;
+          lock->add_waiter(SimpleLock::WAIT_STABLE, new C_Locker_Nudge(this, mdr));
```

Comment:
here allocate a C_Locker_Nudge but not break the loop.   The loop may create lots of C_Locker_Nudge, it's very inefficiency.

I prefer to use another list to track locks that can be nudged at once. just let scatter_tick to process that list later

Reply:
Yes, I'll try to follow your way:-)

Reply:
@ukernel At the point you are refering to, only those locks that are of the rstat flush requests' interest will be added a waiter. If a lock is nudged at once, its rstat_dirty_from would be zero, and it wouldn't be interested by rstat flush requests, so they wouldn't cause a C_Locker_Nudge creation:-)

Reply:
the CInode::need_to_nudge() is controlling whether a nestlock is interested:-)

Reply:
If the 'N' nestlocks (you are interested in) can not be nudged at the moment, you create 'N'  instance of C_Locker_Nudge.  Each one will iterate whole updated_scatterlock list when it get executed. consider that case that 'N' is large and updated_scatterlock list is also large

Reply:
@ukernel I got your poirnt:-) Will fix it:-) Thanks:-)